### PR TITLE
network: several cleanups and fixes for DHCPv4 client

### LINKED ISSUE #28495: `systemd-networkd` automatically adds interface route for DNS while DNS is in separate subnet while there is classless static route available to the DNS subnet
### systemd version the issue has been seen with

253.5

### Used distribution

NixOS 23.05

### Linux kernel version used

6.1.37

### CPU architectures issue was seen on

x86_64

### Component

networkctl

### Expected behaviour you didn't see

No additional interface routes are added.

### Unexpected behaviour you saw

In the specific network, I get the following configuration via DHCP:

- my_ip: `172.0.0.244/24`
- gateway: `172.0.0.1`
- dns1: `10.0.0.153`
- dns2: `10.0.0.153`
- classless, static route: `10.0.0.0/8 via 172.0.0.1`

I confirmed that only this one route is advertised in the DHCP response using WireShark.

Despite that, I get two additional routes:

- `10.0.0.153 dev wlp3s0`
- `10.0.0.154 dev wlp3s0`

These route break the setup: because they are interface routes for the specific target netdev, they are preferred over the classless static route advertised by the DHCP server. Any try to contact the DNS fails, because `wlp3s0` only has a `172.0.0.x/24`, which has no direct way of reaching `10.x.x.x/8`. If I remove the two additional routes manually, my device uses the advertised gateway to reach `10.x.x.x` and everything works.

The respective `.network` file is really boring:

```ini
[Match]
Name=wl* ww*

[Network]
DHCP=yes
LLMNR=false
MulticastDNS=resolve

DNSSEC=no
```

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
Jul 22 18:24:49 zornpad systemd-networkd[57498]: Got message type=method_call sender=:1.341 destination=org.freedesktop.network1 path=/org/freedesktop/network1 interface=org.freedesktop.DBus.Properties member=Get  cookie=2 reply_cookie=0 signature=ss error-name=n/a error-message=n/a
Jul 22 18:24:49 zornpad systemd-networkd[57498]: Sent message type=method_return sender=n/a destination=:1.341 path=n/a interface=n/a member=n/a cookie=56 reply_cookie=2 signature=v error-name=n/a error-message=n/a
Jul 22 18:24:49 zornpad systemd-networkd[57498]: wlp3s0: Flags change: +UP
Jul 22 18:24:49 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=57 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:49 zornpad systemd-networkd[57498]: wlp3s0: Link UP
Jul 22 18:24:49 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received new_station(19) message: bssid=7a:83:c2:cd:19:6d
Jul 22 18:24:50 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received del_station(20) message: bssid=7a:83:c2:cd:19:6d
Jul 22 18:24:50 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received authenticate(37) message.
Jul 22 18:24:50 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received new_station(19) message: bssid=e0:63:da:8f:7a:68
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received authenticate(37) message.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Flags change: +LOWER_UP
Jul 22 18:24:51 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=58 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received associate(38) message.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received connect(46) message: status=0, bssid=e0:63:da:8f:7a:68
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received new_interface(7) message: iftype=station, ssid=dexlab-guest
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Connected WiFi access point: dexlab-guest (e0:63:da:8f:7a:68)
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received control_port_frame_tx_status(139) message.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received control_port_frame_tx_status(139) message.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Received new foreign route (configured): dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: multicast, nexthop: 0, priority: 256, flags: n/a
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Received new foreign route (configured): dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main(254), proto: kernel, type: unicast, nexthop: 0, priority: 256, flags: n/a
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Flags change: +RUNNING
Jul 22 18:24:51 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=59 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Gained carrier
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: found matching network '/etc/systemd/network/95-wireless.network'.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: STARTED on ifindex 4
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: Acquiring DHCPv4 lease.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: LLDP Rx: Started LLDP client
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: nl80211: received notify_cqm(64) message.
Jul 22 18:24:51 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: REQUEST (init-reboot)
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: Received new foreign address (configured): fe80::de41:a9ff:fe38:2df9/64 (valid forever, preferred forever), flags: permanent, scope: link
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: Gained IPv6LL
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: Discovering IPv6 routers
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: NDISC: Started IPv6 Router Solicitation client
Jul 22 18:24:52 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=60 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: Received new foreign route (configured): dst: fe80::de41:a9ff:fe38:2df9/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0, flags: n/a
Jul 22 18:24:52 zornpad systemd-networkd[57498]: wlp3s0: NDISC: Sent Router Solicitation, next solicitation in 4s
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: REBOOTED
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: DISCOVER
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: Received message from RAW socket, processing.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: OFFER
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: REQUEST (requesting)
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: Received message from RAW socket, processing.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: ACK
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: lease expires in 11h 59min 59s
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: T2 expires in 10h 30min
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 client: T1 expires in 6h
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Setting transient hostname: 'zornpad'
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Sent message type=method_call sender=n/a destination=org.freedesktop.hostname1 path=/org/freedesktop/hostname1 interface=org.freedesktop.hostname1 member=SetHostname cookie=61 reply_cookie=0 signature=sb error-name=n/a error-message=n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 address 172.0.0.244/24, gateway 172.0.0.1 acquired from 172.0.0.1
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 address (n/a): 172.0.0.244/24 (valid for 11h 59min 59s, preferred for 11h 59min 59s), flags: tentative, scope: global
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (configured,removing): dst: 172.0.0.1/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (configured,removing): dst: 172.20.0.0/16, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (requesting,configured,removing): dst: 172.0.0.1/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (configured,removing): dst: 10.0.0.0/8, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCP: requested route destination 10.0.0.153/32 is not in the assigned network 172.0.0.0/24, but no gateway is specified, using 'link' scope.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (configured,removing): dst: 10.0.0.153/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCP: requested route destination 10.0.0.154/32 is not in the assigned network 172.0.0.0/24, but no gateway is specified, using 'link' scope.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Requesting DHCPv4 route (configured,removing): dst: 10.0.0.154/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: State changed: configured -> configuring
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=62 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: link_check_ready(): address 172.0.0.244/24 is not ready.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 address (requesting): 172.0.0.244/24 (valid for 11h 59min 59s, preferred for 11h 59min 59s), flags: tentative, scope: global
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 route (requesting,configured,removing): dst: 172.0.0.1/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 route (requesting,configured,removing): dst: 172.20.0.0/16, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 route (requesting,configured,removing): dst: 10.0.0.0/8, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 route (requesting,configured,removing): dst: 10.0.0.153/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Configuring DHCPv4 route (requesting,configured,removing): dst: 10.0.0.154/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received updated DHCPv4 address (configured): 172.0.0.244/24 (valid for 11h 59min 59s, preferred for 11h 59min 59s), flags: n/a, scope: global
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=63 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: link_check_ready(): IPv4LL:no DHCPv4:no DHCPv6:no DHCP-PD:no NDisc:yes
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: State changed: configuring -> configured
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=64 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received new foreign route (configured): dst: 172.0.0.244/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: host, table: local(255), proto: kernel, type: local, nexthop: 0, priority: 0, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered foreign route (configured): dst: 172.0.0.255/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: local(255), proto: kernel, type: broadcast, nexthop: 0, priority: 0, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered foreign route (configured): dst: 172.0.0.0/24, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: kernel, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: dhcp4_check_ready(): DHCPv4 address and routes are not set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered DHCPv4 route (configured): dst: 172.0.0.1/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: dhcp4_check_ready(): DHCPv4 address and routes are not set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered DHCPv4 route (configured): dst: 172.20.0.0/16, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: dhcp4_check_ready(): DHCPv4 address and routes are not set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered DHCPv4 route (configured): dst: 10.0.0.0/8, src: n/a, gw: 172.0.0.1, prefsrc: 172.0.0.244, scope: global, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: dhcp4_check_ready(): DHCPv4 address and routes are not set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered DHCPv4 route (configured): dst: 10.0.0.153/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: dhcp4_check_ready(): DHCPv4 address and routes are not set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: Received remembered DHCPv4 route (configured): dst: 10.0.0.154/32, src: n/a, gw: n/a, prefsrc: 172.0.0.244, scope: link, table: main(254), proto: dhcp, type: unicast, nexthop: 0, priority: 1024, flags: n/a
Jul 22 18:24:54 zornpad systemd-networkd[57498]: wlp3s0: DHCPv4 address and routes set.
Jul 22 18:24:54 zornpad systemd-networkd[57498]: Got message type=method_return sender=:1.338 destination=:1.329 path=n/a interface=n/a member=n/a  cookie=15 reply_cookie=61 signature=n/a error-name=n/a error-message=n/a
Jul 22 18:24:57 zornpad systemd-networkd[57498]: wlp3s0: NDISC: Sent Router Solicitation, next solicitation in 8s
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #28495.

### CODE REVIEW

File: src/network/networkd-dhcp4.c [hunk 1]
```diff
@@ -52,6 +52,39 @@ void network_adjust_dhcp4(Network *network) {
                 network->dhcp_client_identifier = network->dhcp_anonymize ? DHCP_CLIENT_ID_MAC : DHCP_CLIENT_ID_DUID;
 }
 
+static int dhcp4_get_static_routes(Link *link, sd_dhcp_route ***ret_routes, size_t *ret_num) {
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: This returns a list of routes from either sd_dhcp_lease_get_classless_routes or  sd_dhcp_lease_get_static_routes. So we call the first set first "classless routes" and then "static classless routes", and then merge that with the second set "static routes" into "static routes". This is very confusing. Please keep the naming consistent with the `sd_dhcp_lease_get_*` functions.

File: src/network/networkd-dhcp4.c [hunk 2]
```diff
@@ -52,6 +52,39 @@ void network_adjust_dhcp4(Network *network) {
...
+        _cleanup_free_ sd_dhcp_route **routes = NULL;
+        int r;
+
+        assert(link);
+        assert(link->dhcp_lease);
+
+        /* if the DHCP server returns both a Classless Static Routes option and a Static Routes option,
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c]: "If"

File: src/network/networkd-dhcp4.c [hunk 3]
```diff
@@ -130,6 +130,97 @@ static int dhcp4_get_static_routes(Link *link, sd_dhcp_route ***ret_routes, size
         return 0; /* static */
 }
```
Comment (keszybz) [inline src/network/networkd-dhcp4.c:136]: > If not found, skip to configure the route.

"skip to" is used with a target. E.g. "skip to the end". If you want to skip *over* something, you don't use "to". E.g. "skip over the middle", "skip the middle".

"If not found, do not configure the route." or "If not found, skip configuration of the route.".

**Comment by keszybz** [comment]
I set `reviewed/needs-rework` because I think somebody else should take a look if there's another version.

File: src/network/networkd-dhcp4.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-dhcp4.c]: More precisely, classless static routes + static routes into static routes. haha.
`dhcp4_get_classless_static_or_static_routes()` ?? That sounds slightly long and redundant for me, but ok.

**Comment by yuwata** [comment]
@keszybz Thank you for the review.
- renamed function,
- added more comments in the code, 
- extended commit messages.

PTAL.

**Review by bluca — APPROVED** [review]

**Pull request merged**